tests/test_jit.py

```diff
@@ -420,5 +317,114 @@ def f(x, y):
         return x + y
 
     g = jtu.Partial(f, jrandom.normal(getkey(), ()))
-    fn = jtu.Partial(f, True)
-    eqx.filter_jit(g, fn=fn)
+    eqx.filter_jit(g)
+
+
+def test_buffer_donation(getkey):
+    def f(x):
+        return x + 1
+
+    def _test(jit_fun, x):
+        old_p = x.unsafe_buffer_pointer()
+        new_x = jit_fun(x)
+        assert new_x.unsafe_buffer_pointer() == old_p
+        assert x.is_deleted()
+        return new_x
+
+    f_jit = eqx.filter_jit(f)
+
+    assert _eq(_test(f_jit, jnp.array(0)), jnp.array(1.0))
```

Comment:
The `_eq` function should probably be removed and `helpers.shaped_allclose` used instead. (Which is basically the same and handles pytrees and floating-point arrays.)